### ISSUE TITLE
Allow to run queries through LiveQueryClient

### ISSUE THREAD

**Issue by messagenius-admin** [parse-community/parse-server#9086]
### Current Limitation

When developing applications using Parse, you want to sync the App with the Server when the App is re-opened after being killed, shut down or restored from the background. 

You need a Query to obtain the most current data and subscribe to a LiveQuery to receive real-time updates while the application is in use. 

In complex apps with multiple collections, a large number of Query + LiveQuery combinations are necessary every time the app is restored from the background, requiring several REST calls and a WebSocket connection.

### Feature / Enhancement Description

It should be possible to run Queries through a LiveQueryClient, and/or to request query results along with LiveQuery subscription to enable syncing collections and subscribing to LiveQueries with a single WebSocket connection.

We believe that this improvement makes a difference. It's easy to develop and maintain, with minimal changes to the Parse Server and client SDKs. 

This is not a breaking change, it is fully retrocompatible, and existing apps can be improved easily.

**EDIT Apr 30:**
It would be beneficial to have the option to use WebSocket for all requests, including Queries, Saves, Cloud Functions, and all CRUD operations.

### Example Use Case

Imagine a complex TODO mobile app with **six** collections: boards, tasks, goals, categories, tags,  and users.

Every time the user opens the App after from background, especially after being offline, you need to run **six** Parse Queries  and subscribe to them.

Each of the **six**  queries is an HTTP/REST API call, while all LiveQueries travel on a single Websocket connection.

**Current Approach:**

```JavaScript
const tasks = getTasksFromLocalDb();

const query = new Parse.Query('Tasks');
query.greaterThan("updatedAt", getLastTimeOpenedApp());

// This is the standard (and only) way so far to retrieve objects. 
const results = await query.find(); // For each Query, the SDK triggers the REST API Call we want to avoid.
results.forEach(task => tasks[task.id] = task);

const subscription = await query.subscribe();
subscription.on('create', task => tasks[task.id] = task);

```

**Possible Alternative 1:**
Get query results leveraging the subscription

```JavaScript
subscription.on('result', task => tasks[task.id] = task);
subscription.find(); // or subscription.first(); or getResults() or runQuery(), or any other naming that makes sense.
```

**Possible Alternative 2:**
Allow running any query from a LiveQueryClient. This potentially enables developers to run queries through WebSocket, detached from a subscription.

```JavaScript
const client = new LiveQueryClient({ ... });
const querySubscription = client.querySubscribe(query);
querySubscription.on('result', task => tasks[task.id] = task);
```

Ideally, both alternatives 1 and 2 should be developed, with one serving as a shortcut for the other.

### 3rd Party References

Firebase `onSnapshot` event.

**Comment by mtrezza** [comment]
I suppose this feature requires changes in Parse Server and at least the Parse JS SDK to a Parse client SDK to test with?

**Comment by messagenius-admin** [comment]
Created a JS SDK Issue ere: https://github.com/parse-community/Parse-SDK-JS/issues/2114

**Comment by mtrezza** [comment]
Note: the scope of this issue in regard to the bounty includes https://github.com/parse-community/Parse-SDK-JS/issues/2114

**Comment by mortenmo** [comment]
Is this trying to solve a performance problem with too many https calls? Working a lot with ParseJS trying to make it work great offline lately, in my humble opinion this is probably not worth the effort. 

Its not like you have to write the code for each object type but create a generic store that executes the logic you describe. We built one loosely based on `https://github.com/parse-community/parse-react/blame/master/packages/parse-react-base/src/useParseQuery.ts` which has a decent way of doing what you describe. I do think it is useful to get the "current results" when you subscribe (because you almost always need that), but this could be solved in the JS SDK completely today without doing pure queries over the websocket.

If it is a performance challenge to fix, why not put _every_ request over the livestream, not just one you subscribe to. Then I'd like a mode where just all server comms go over the websocket and never uses https (ad-hoc finds, saves, etc).

**Comment by messagenius-admin** [comment]
Hi @mortenmo,
thanks for sharing.

In this particular Issue, I'm trying to solve performance issues but with minimal impact on existing code and this is just a piece of a larger improvement.

The problem is that, even with the generic handler that you implemented in React, the client has to run a query (which means an HTTP call) each time the app is restored from an offline period, (even after being in background for just 1 second, in iOS)


The code you pasted:
> ... but create a generic store that executes the logic you describe. We built one loosely based on `https://github.com/parse-community/parse-react/blame/master/packages/parse-react-base/src/useParseQuery.ts`
the client is doing 3 operations for each Class: 

 is a great example of how each client SDK should manage to sync; in fact, we would need it for iOS, Android, Flutter, and others. We should open a feature request for each client.

 
> why not put _every_ request over the live stream [...] all server comms go over the websocket [...] (ad-hoc finds, saves, etc)

I would love it, but it may be a longer shot. I'll edit the Issue to add this option.

**Comment by paulovitin** [comment]
Hi @messagenius-admin @mtrezza — I'm looking at implementing this on the server side and would like to align on a few protocol decisions before writing code, since the shape we pick here is what every client SDK (JS, iOS, Android, Flutter) will have to follow.

My current plan is to ship this in three small PRs:
1. Refactor: extract the auth/CLP/protected-fields/regex/complexity pipeline out of `_handleSubscribe` into a reusable `_authorizeQuery` helper (no behavior change).
2. Add a new `op: 'query'` operation to the LiveQuery WebSocket protocol — runs a query via `rest.find` using the reused authorization pipeline, responds with results, creates no subscription (covers Alternative 2 from the issue).
3. Add a `fetch: true` flag on the existing `subscribe` op that emits an initial snapshot right after `subscribed` (covers Alternative 1, implemented as a shortcut over #2).

Four questions before I start:

**1. Protocol naming.** The issue floats several names (`find`, `getResults`, `runQuery`, `result`, `snapshot`). What do you prefer for:
   - the request op: `query` vs `find`?
   - the response op: `result` vs `snapshot` vs `results`?

**2. Triggers.** For the standalone `op: 'query'` (no subscription), should it fire `beforeSubscribe`, or only `beforeFind` (which `rest.find` already dispatches internally)? My inclination is **only `beforeFind`** — it's semantically a query, not a subscription — but I want to confirm, since `beforeSubscribe` has historically been the security boundary for LiveQuery traffic.

**3. Snapshot ordering / race window.** When `subscribe` is called with `fetch: true`, there's a small race window between the snapshot being taken and live events starting to flow — an object created in that window could be missed, or delivered twice. Two options:
   - **Server-side ordering**: the server takes the snapshot *before* activating the subscription and guarantees no events are dropped (more work on the server, cleaner contract).
   - **Client-side dedup**: the server emits `subscribed` → `result` and clients dedupe by `objectId` (simpler server, every SDK has to implement dedup).
   
   Which do you prefer?

**4. Pagination / count in v1.** Should the standalone `op: 'query'` support `count: true`, `limit`, `skip`, `order`, `include` from day one (matching REST `find` parity), or start minimal and add them later?

Also — should I tag the sponsor of the bounty on these decisions, or is mtrezza the right contact for protocol shape?

Happy to start on PR 1 (the refactor) in parallel since it's behavior-preserving and useful either way.

**Comment by simonegiacco** [comment]
Hey @paulovitin — great to see this moving forward, and these are exactly the right questions to align on before touching the SDKs.

**1. Protocol naming**

Taking inspiration from the existing JS SDK convention:

```js
const query = new Parse.Query(GameScore);
const results = await query.find();
```

I'd mirror that symmetry in the protocol: `op: find` on the request side, `op: results` on the response. It'll feel immediately intuitive to anyone already familiar with the SDK surface.

For a v2-ready design, we could extend this naturally to `op: query.find`, `op: query.first`, `op: query.count` on the request side — with corresponding responses `op: results`, `op: result`, and `op: count`. Keeps the namespace clean and extensible.

**2. Triggers**

Agreed — only `beforeFind` for the standalone `op: find`. 

For `subscribe` with `fetch: true`, both `beforeFind` and `beforeSubscribe` fire implicitly, which is the right behavior.

**3. Snapshot ordering / race window**

I'd go with client-side dedup. Query results should be treated as *updates* rather than *creates*, which means deduplication by `objectId` is already necessary on the client regardless — it's not extra work imposed by this design choice, it's just sound LiveQuery handling. 

**4. Pagination / count in v1**

Full parity from day one gets my vote — it significantly lowers the barrier to adoption. 
In practice, personally, we always use `limit` and `include`; `-updatedAt` ordering is our default. `count` is occasionally useful. `include` is the one we'd miss most if deferred, but we could live without it initially if it simplifies the first cut.

---

More broadly: if we design this cleanly, WebSocket could become the default transport for every Parse operation — queries, cloud functions, the works — with REST as the fallback.

**Comment by mtrezza** [comment]
@simonegiacco Mind that there is an existing PR https://github.com/parse-community/parse-server/pull/9864 which seems to be near completion, but is stale due to lack of response from the issue author, tagged in https://github.com/parse-community/parse-server/pull/9864#issuecomment-3482608325.

**Comment by Corvus400** [comment]
@mtrezza I opened #10470 as a continuation of #9864 for this issue.

The PR is ready for review and currently waiting for maintainer feedback. When you have time, could you please take a look and let me know whether this direction works for Parse Server?

If there is anything you would prefer to see changed before review, I would be glad to adjust it.

**Comment by Corvus400** [comment]
Hi @mtrezza,

PR #10470 has completed review, all CI checks are passing, and GitHub reports it as approved and mergeable. It is ready to merge whenever convenient.

Thank you for taking a look.

**Comment by mtrezza** [comment]
@Corvus400 This feature has a sponsored bounty, so we'll see if @simonegiacco @messagenius-admin have any feedback.
